helpers/location_service.py

```diff
@@ -1,166 +1,240 @@
-        if message['type'] == "state":
-            print("Message received: %s" % message)
-            SIGNALS.locs_recieved.emit(message['locations'],
-                                       message.get('waypoints', {}))
-
-    def _on_error(self, ws, error):
-        print("Connection error: %s" % error)
-
-    def _on_open(self, ws):
-        print("Connection opened.")
-
-    def _on_close(self, ws, status_code, close_message):
-        print(f"Connection closed: ({status_code}) {close_message}")
+    # Websocket message handler - handles any incoming messages from the websocket server
+    def message(self, message):
+        QApplication.instance()._signals["locationsharing"].textMessageReceived.emit(message)
+
+    # Parses messages from the websocket server
+    def parse(self, websocket_message):
+        message = loads(websocket_message)
+        if message["type"] == "state":
+            # Only process locations if there is data for our location
+            if message.get("locations", False).get(self.zone_name.lower(), False):
+                # Interate through players in the zone
+                for player in message["locations"][self.zone_name.lower()]:
+                    #Process any players that are not us
+                    if player.lower() != self.character_name.lower():
+                        p_data = message["locations"][self.zone_name.lower()][player]
+                        p_timestamp = datetime.fromisoformat(str(p_data["timestamp"]))
+                        p_point = MapPoint(x=int(p_data["x"]), y=int(p_data["y"]), z=int(p_data["z"]))
+                        #Add the player map point to the map
+                        QApplication.instance()._parsers_dict["maps"]._map.add_player(player, p_timestamp, p_point)
+
+                # Remove players that aren't in the zone
+                players_to_remove = []
+                # Check players in the currently loaded QT map
+                for player in QApplication.instance()._parsers_dict["maps"]._map._data.players:
+                    # if the player is no longre in the zone and not ourselves, append them to be removed
+                    if player not in message["locations"][self.zone_name.lower()] and player != "__you__":
+                        players_to_remove.append(player)
+                for player in players_to_remove:
+                    QApplication.instance()._parsers_dict["maps"]._map.remove_player(player)
+
+            # Only process waypoints if there is data for our location
+            if message.get("waypoints", False).get(self.zone_name.lower(), False):
+                # Iterate through the waypoints in the zone
+                for waypoint in message["waypoints"][self.zone_name.lower()]:
+                    w_data = message["waypoints"][self.zone_name.lower()][waypoint]
+                    w_point = MapPoint(x=int(w_data["x"]), y=int(w_data["y"]), z=int(w_data["z"]))
+                    w_icon = w_data.get("icon", "corpse")
```

Comment:
The server-side waypoints WILL have an icon field set for the single type we currently send (corpse), but there will probably be others... so maybe we default to something more generic than corpse? 😅